### PR TITLE
AppBar can now have custom titleHeight

### PR DESCRIPTION
**Pull request by RaviKavaiya**
Allows AppBar (and SliverAppBar) to have any height we specify.

## Description

Previously, AppBar's height was limited to 56.0 using a constant `kToolbarHeight`. This allows us to change AppBar's height using `titleHeight` parameter. If we omit this parameter, it fallbacks to `kToolbarHeight` (the default behavior), otherwise acts accordingly.

## Related Issues

#23373 solved here
#7330 looks related

### EDIT: About second commit:
- When we specify `titleHeight`, the icons (`leading` and `actions`) are also centered. So, this commit introduces a boolean `centerIcons`, which when set to `false`, the icons stay on top.
- Also fixes some analyzer warnings

## Tests

I added the following tests:

- Added test in `app_bar_test.dart` so that we can check custom title height provided.
- Added test in `floating_action_button_location_test.dart` so that we can check mini-start top FAB location.
- Added test in `scrollable_semantics_test.dart` so that showOnScreen works with pinned app bar.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/material/app_bar.dart [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
 import 'dart:math' as math;
 
+import 'package:flutter/cupertino.dart';
```
Comment (dnfield) [inline packages/flutter/lib/src/material/app_bar.dart]: We should not be depending on cupertino in a material widget.

File: packages/flutter/lib/src/material/app_bar.dart [hunk 2]
```diff
@@ -23,20 +24,23 @@ import 'tabs.dart';
 import 'text_theme.dart';
 import 'theme.dart';
 
-const double _kLeadingWidth = kToolbarHeight; // So the leading button is square.
+const double _kLeadingWidth =
+    kToolbarHeight; // So the leading button is square.
```
Comment (dnfield) [inline packages/flutter/lib/src/material/app_bar.dart]: flutter style guide says to keep these on one line, even if htye go over 80 chars.

File: packages/flutter/lib/src/material/app_bar.dart [hunk 3]
```diff
@@ -45,7 +49,8 @@ class _ToolbarContainerLayout extends SingleChildLayoutDelegate {
   }
 
   @override
-  bool shouldRelayout(_ToolbarContainerLayout oldDelegate) => false;
+  bool shouldRelayout(_ToolbarContainerLayout oldDelegate) =>
```
Comment (dnfield) [inline packages/flutter/lib/src/material/app_bar.dart]: This and the next line should be on one line.

File: packages/flutter/lib/src/material/app_bar.dart [hunk 4]
```diff
@@ -194,16 +199,20 @@ class AppBar extends StatefulWidget implements PreferredSizeWidget {
     this.primary = true,
     this.centerTitle,
     this.titleSpacing = NavigationToolbar.kMiddleSpacing,
+    this.titleHeight,
```
Comment (dnfield) [inline packages/flutter/lib/src/material/app_bar.dart]: Default this to `kToolbarHeight` and assert that it is not null.

File: packages/flutter/lib/src/material/app_bar.dart [hunk 5]
```diff
@@ -359,6 +368,21 @@ class AppBar extends StatefulWidget implements PreferredSizeWidget {
   /// Defaults to [NavigationToolbar.kMiddleSpacing].
   final double titleSpacing;
 
+  /// The amount of height to be used for the [title].
+  ///
+  /// If not provided, then defaults to [kToolbarHeight].
```
Comment (dnfield) [inline packages/flutter/lib/src/material/app_bar.dart]: Maybe just `The height used for the [title].` and then `This value must not be null, and defaults to [kToolbarHeight].`

File: packages/flutter/lib/src/material/app_bar.dart [hunk 6]
```diff
@@ -359,6 +368,21 @@ class AppBar extends StatefulWidget implements PreferredSizeWidget {
...
+  final double titleHeight;
+
+  /// Specifies whether the [AppBar.leading] and [AppBar.actions] should always be placed vertically center of the AppBar.
```
Comment (dnfield) [inline packages/flutter/lib/src/material/app_bar.dart]: nit: doc strings should be wrapped at 80 chars.

not nit: I think you mean horizontally, right?

also not nit: why are we adding this property?  I think we should avoid adding it until we know someone wants it.

File: packages/flutter/lib/src/material/app_bar.dart [hunk 7]
```diff
@@ -377,16 +401,17 @@ class AppBar extends StatefulWidget implements PreferredSizeWidget {
   /// bar is scrolled.
   final double bottomOpacity;
 
-  /// A size whose height is the sum of [kToolbarHeight] and the [bottom] widget's
+  /// A size whose height is the sum of [kToolbarHeight] (or [titleHeight], if given) and the [bottom] widget's
```
Comment (dnfield) [inline packages/flutter/lib/src/material/app_bar.dart]: I think this should just be `A size whose height is the sum of [titleHeight] and the [bottom] widget's preferred height`, assuming we change title height to never be null.

File: packages/flutter/lib/src/material/app_bar.dart [hunk 8]
```diff
@@ -426,33 +451,35 @@ class _AppBarState extends State<AppBar> {
...
-      ?? appBarTheme.actionsIconTheme
-      ?? overallIconTheme;
-    TextStyle centerStyle = widget.textTheme?.title
-      ?? appBarTheme.textTheme?.title
-      ?? theme.primaryTextTheme.title;
-    TextStyle sideStyle = widget.textTheme?.body1
-      ?? appBarTheme.textTheme?.body1
-      ?? theme.primaryTextTheme.body1;
+    final bool useCloseButton =
+        parentRoute is PageRoute<dynamic> && parentRoute.fullscreenDialog;
```
Comment (dnfield) [inline packages/flutter/lib/src/material/app_bar.dart]: Please revert formatting changes here and below.

Comment (dnfield) [reply packages/flutter/lib/src/material/app_bar.dart]: (We do not use dartfmt in this repo)

File: packages/flutter/lib/src/material/app_bar.dart [hunk 9]
```diff
@@ -528,13 +555,14 @@ class _AppBarState extends State<AppBar> {
       trailing: actions,
       centerMiddle: widget._getEffectiveCenterTitle(theme),
       middleSpacing: widget.titleSpacing,
+      centerIcons: widget.centerIcons,
     );
 
-    // If the toolbar is allocated less than kToolbarHeight make it
+    // If the toolbar is allocated less than kToolbarHeight (or titleHeight, if given)  make it
```
Comment (dnfield) [inline packages/flutter/lib/src/material/app_bar.dart]: ditto on docstring from above

File: packages/flutter/lib/src/material/app_bar.dart [hunk 10]
```diff
@@ -550,15 +578,17 @@ class _AppBarState extends State<AppBar> {
         children: <Widget>[
           Flexible(
             child: ConstrainedBox(
-              constraints: const BoxConstraints(maxHeight: kToolbarHeight),
+              constraints: BoxConstraints(
+                  maxHeight: widget.titleHeight ?? kToolbarHeight),
```
Comment (dnfield) [inline packages/flutter/lib/src/material/app_bar.dart]: We should just make titleHeight non-null and drop the ?? here.

File: packages/flutter/lib/src/material/app_bar.dart [hunk 11]
```diff
@@ -550,15 +578,17 @@ class _AppBarState extends State<AppBar> {
...
               child: appBar,
             ),
           ),
           if (widget.bottomOpacity == 1.0)
             widget.bottom
           else
             Opacity(
-              opacity: const Interval(0.25, 1.0, curve: Curves.fastOutSlowIn).transform(widget.bottomOpacity),
+              opacity: const Interval(0.25, 1.0, curve: Curves.fastOutSlowIn)
+                  .transform(widget.bottomOpacity),
```
Comment (dnfield) [inline packages/flutter/lib/src/material/app_bar.dart]: revert formatting

**Review by dnfield — CHANGES_REQUESTED** [review]
I didn't quite make it all the way through - if you could revert all the dartfmt changes it would help make the review easier :)

**Comment by RaviKavaiya** [comment]
@dnfield I agree about the dartfmt issue. I am also facing this issue. All the files in this repo are formatted in different manner. 
I use Android Studio. Can you give me that Code Style settings file (if you use ine), so that I can also format my code accordingly?

**Comment by RaviKavaiya** [comment]
@dnfield Thanks for your support. 
I am *closing* this Pull Request, as I am facing many formatting issues.
I will create a new Pull Request in couple of days(with the same details).

**Comment by dnfield** [comment]
We format by hand in this repo - see https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#formatting

**Pull request closed without merging**